File: src/service/growth.ts
/**
 * 世界 / 实验室 / 成长域 —— `backend/service/growth.py` 的 TypeScript 移植。
 *
 * **P2 阶段的定位**：契约里这些端点必须存在且形状正确，但它们的游戏规则
 * （材料掉落、建筑解锁、徽章、布局）属于 P5 成长体系。因此本模块只实现
 * "从学习事实中可推导"的部分，规则常量集中放在文件顶部，方便 P5 整体替换。
 *
 * 实验室（§7）不在本模块维护规则：解锁判定统一走 `src/lab` +
 * `config/lab/v0.yaml`（构建产物），这里只提供 `labPayload`（内含
 * 派生等级 level_of，ADR-0002，读时现算不落库）。
 *
 * 推导原则：游戏状态永远不是学习状态的权威（见 db/migrations 的注释）。
 * 这里的材料 / 建筑 / 徽章只读 inventory / unlock / attempt / proficiency_state。
 *
 * session 参数的对应物：Python 所有函数第一个参数都是 SQLAlchemy session，
 * 事务边界由路由层控制。TS 侧只读函数内部直接用全局 `getDb()`；唯一会
 * 写库的 `buildPayload` 接收路由层开好的事务（见 DbExecutor 的说明）。
 */
import { and, asc, eq, inArray } from "drizzle-orm";

import type { ContentBundle } from "@/src/content/types";
import { challengeBeats } from "@/src/content/types";
import { getDb, type DbExecutor, type DbTx } from "@/src/db/client";
import { attempt, inventory, unlock } from "@/src/db/schema";
import type { AlgorithmConfig } from "@/src/engine/config";
import type { CompetencyGraph } from "@/src/engine/graph";
import { deriveLevel, isMastered } from "@/src/engine/state-machine";
import type { ChildLearningState } from "@/src/engine/types";
import { listExperiments } from "@/src/lab/service";
import { pyRound } from "@/src/py/pyround";
import { sortedBy, sortedStrings } from "@/src/py/pysort";

// ── P5 待接管的常量（内容是资产，规则应可配置） ─────────────

export interface UniverseDef {
  code: string;
  name: string;
  emoji: string;
}

export interface BuildingDef {
  code: string;
  name: string;
  emoji: string;
  cost: Record<string, number>;
  unlocks: string[];
}

export interface BadgeDef {
  code: string;
  name: string;
  emoji: string;
  competency: string;
}

export const UNIVERSES: UniverseDef[] = [
  { code: "number_station", name: "数字车站", emoji: "🚂" },
  { code: "detective", name: "侦探社", emoji: "🦊" },
];

export const BUILDINGS: BuildingDef[] = [
  {
    code: "ticket_booth",
    name: "售票亭",
    emoji: "🎫",
    cost: { wood: 5, coin: 2 },
    unlocks: [],
  },
  {
    code: "platform",
    name: "站台",
    emoji: "🛤️",
    cost: { wood: 10, coin: 5 },
    unlocks: ["universe.detective"],
  },
];

export const BADGES: BadgeDef[] = [
  { code: "first_make_ten", name: "第一次凑十", emoji: "🎖️", competency: "make_ten" },
  { code: "first_carry", name: "第一次进位", emoji: "🚃", competency: "carry_add" },
];

export const MATERIAL_CODES = ["wood", "coin", "gem", "seed"];

export const COMPETENCY_EMOJI: Record<string, string> = {
  sd_add_10: "➕",
  sd_sub_10: "➖",
  sd_add_20: "🔢",
  sd_sub_20: "🔽",
  make_ten: "🔟",
  place_value: "🚃",
  td_add_nocarry: "🧮",
  td_sub_nocarry: "📉",
  carry_add: "🔁",
  borrow_sub: "🔄",
};

export const DEFAULT_UNIVERSE = "number_station";

// ── 基础读取 ───────────────────────────────────────────────

export async function materialsOf(
  childId: number,
  client?: DbExecutor,
): Promise<Record<string, number>> {
  const db = client ?? getDb();
  const out: Record<string, number> = {};
  for (const code of MATERIAL_CODES) {
    out[code] = 0;
  }
  const rows = await db.select().from(inventory).where(eq(inventory.childId, childId));
  for (const row of rows) {
    // Python：`out[row.item_code] = int(row.count or 0)` —— item_code 不在
    // MATERIAL_CODES 里也会加键（未来掉落物不因常量表滞后而丢失）
    out[row.itemCode] = Math.trunc(row.count ?? 0);
  }
  return out;
}

export async function unlockedCodes(childId: number, client?: DbExecutor): Promise<Set<string>> {
  const db = client ?? getDb();
  const rows = await db
    .select({ unlockCode: unlock.unlockCode })
    .from(unlock)
    .where(eq(unlock.childId, childId));
  return new Set(rows.map((row) => row.unlockCode));
}

export function competencyEmoji(code: string): string {
  return COMPETENCY_EMOJI[code] ?? "📘";
}

/** 契约 §7：实验列表。解锁规则整套由 `src/lab` 负责（配置构建产物）。 */
export function labPayload(state: ChildLearningState, cfg: AlgorithmConfig): Record<string, unknown> {
  const levelOf = (competencyCode: string): string => {
    const signals = state.competencies.get(competencyCode);
    if (signals === undefined) {
      return cfg.level_order[0]!;
    }
    return deriveLevel(signals, cfg);
  };
  return {
    experiments: listExperiments(levelOf, [...cfg.level_order]),
  };
}

// ── 故事完成度（用 attempt 的 slot_code 反推） ─────────────

export async function completedStoryCodes(childId: number, bundle: ContentBundle): Promise<string[]> {
  const db = getDb();
  const done: string[] = [];
  for (const story of bundle.stories.values()) {
    const slotCodes = challengeBeats(story)
      .map((beat) => beat.slot_code)
      .filter((code): code is string => Boolean(code));
    if (slotCodes.length === 0) {
      continue;
    }
    const rows = await db
      .selectDistinct({ slotCode: attempt.slotCode })
      .from(attempt)
      .where(and(eq(attempt.childId, childId), inArray(attempt.slotCode, slotCodes)));
    const hit = new Set<string>();
    for (const row of rows) {
      if (row.slotCode) {
        hit.add(row.slotCode);
      }
    }
    if (slotCodes.every((code) => hit.has(code))) {
      done.push(story.code);
    }
  }
  return done;
}

// ── GET /v1/world ──────────────────────────────────────────

export async function worldPayload(
  child: { id: number; name: string },
  state: ChildLearningState,
  cfg: AlgorithmConfig,
  bundle: ContentBundle,
  budgetMinutes: number,
): Promise<Record<string, unknown>> {
  const completed = new Set(await completedStoryCodes(child.id, bundle));
  const stories = sortedBy([...bundle.stories.values()], (s) => s.order_index, (s) => s.code);

  const universes: Record<string, unknown>[] = [];
  for (const universe of UNIVERSES) {
    const universeStories = stories.filter((s) => s.universe === universe.code);
    const total = universeStories.length;
    const doneCount = universeStories.filter((s) => completed.has(s.code)).length;
    const progress = total ? doneCount / total : 0.0;
    universes.push({
      code: universe.code,
      name: universe.name,
      emoji: universe.emoji,
      unlocked:
        universe.code === DEFAULT_UNIVERSE ||
        (await unlockedCodes(child.id)).has(`universe.${universe.code}`),
      // `round(progress, 4)` 是 Python 的 half-even，用 pyRound
      progress: pyRound(progress, 4),
      stories: universeStories.map((story) => ({
        code: story.code,
        title: story.title,
        completed: completed.has(story.code),
        unlocked: true,
      })),
    });
  }

  const todayStory = stories.find((s) => !completed.has(s.code)) ?? null;
  const today =
    todayStory !== null
      ? {
          headline: todayStory.title,
          subtitle: todayStory.summary || "去数字车站帮它一把",
          story_code: todayStory.code,
          universe_code: todayStory.universe,
          estimated_minutes: todayStory.duration_min || budgetMinutes,
          completed: false,
        }
      : {
          // 没有故事内容（P2 未接入）时的降级：形状不变，字段指向默认宇宙
          headline: "今天来练一练数字吧",
          subtitle: "去数字车站练一练",
          story_code: null,
          universe_code: DEFAULT_UNIVERSE,
          estimated_minutes: budgetMinutes,
          completed: false,
        };

  const codes = await unlockedCodes(child.id);
  return {
    child: { id: child.id, name: child.name },
    today,
    universes,
    lab_unlocked: true,
    detective_unlocked: codes.has("universe.detective"),
    growth_summary: {
      materials: await materialsOf(child.id),
      buildings: sortedStrings([...codes])
        .filter((code) => code.startsWith("building."))
        .map((code) => code.slice(code.indexOf(".") + 1)),
      newest_badge: await newestBadge(child.id, state),
    },
  };
}

// ── GET /v1/growth ─────────────────────────────────────────

export async function badgePayloads(
  childId: number,
  state: ChildLearningState,
): Promise<Record<string, unknown>[]> {
  const db = getDb();
  const out: Record<string, unknown>[] = [];
  for (const badge of BADGES) {
    const signals = state.competencies.get(badge.competency);
    const earned = signals !== undefined && signals.sample_count > 0;
    let earnedAt: string | null = null;
    if (earned) {
      const rows = await db
        .select({ createdAt: attempt.createdAt })
        .from(attempt)
        .where(and(eq(attempt.childId, childId), eq(attempt.competencyCode, badge.competency)))
        .orderBy(asc(attempt.seq))
        .limit(1);
      // created_at 列 notNull，Python 的 `row[0] is not None` 检查在这里
      // 只剩"有没有行"一个分支
      if (rows.length > 0) {
        earnedAt = iso(rows[0]!.createdAt);
      }
    }
    out.push({
      code: badge.code,
      name: badge.name,
      emoji: badge.emoji,
      earned,
      earned_at: earnedAt,
    });
  }
  return out;
}

async function newestBadge(
  childId: number,
  state: ChildLearningState,
): Promise<Record<string, unknown> | null> {
  const earned = (await badgePayloads(childId, state)).filter((b) => b["earned"]);
  if (earned.length === 0) {
    return null;
  }
  const last = earned[earned.length - 1]!;
  return {
    code: last["code"],
    name: last["name"],
    emoji: last["emoji"],
  };
}

export async function growthPayload(
  child: { id: number; name: string },
  state: ChildLearningState,
  cfg: AlgorithmConfig,
  bundle: ContentBundle,
  graph: CompetencyGraph,
): Promise<Record<string, unknown>> {
  const order = graph.topologicalOrder();
  const positionIndex = new Map(order.map((code, index) => [code, index]));

  const nodes: Record<string, unknown>[] = [];
  for (const code of order) {
    const competency = bundle.competencies.get(code);
    if (competency === undefined) {
      continue;
    }
    const signals = state.competencies.get(code);
    const level = signals !== undefined ? deriveLevel(signals, cfg) : cfg.level_order[0]!;
    nodes.push({
      code,
      name: competency.name,
      level,
      level_label: cfg.levelLabel(level),
      mastered: isMastered(state, code, graph, cfg),
      emoji: competencyEmoji(code),
      unlocked: true,
      position: { x: positionIndex.get(code) ?? 0, y: competency.stage },
    });
  }

  const edges: Record<string, string>[] = [];
  for (const code of order) {
    for (const prereq of graph.prerequisites(code)) {
      edges.push({ from: prereq, to: code });
    }
  }

  const built = new Set(
    [...(await unlockedCodes(child.id))]
      .filter((code) => code.startsWith("building."))
      .map((code) => code.slice(code.indexOf(".") + 1)),
  );
  return {
    tree: { nodes, edges },
    materials: await materialsOf(child.id),
    buildings: BUILDINGS.map((row) => ({
      code: row.code,
      name: row.name,
      emoji: row.emoji,
      built: built.has(row.code),
      cost: row.cost,
    })),
    badges: await badgePayloads(child.id, state),
  };
}

/**
 * POST /v1/growth/build：材料足够则扣减并解锁。
 *
 * 只在路由层开好的事务里跑（Python 侧 `session.commit()` 也在路由层）。
 * 材料不足**不是错误**：返回 200 + `built: false` + 原因。
 */
export async function buildPayload(
  tx: DbTx,
  childId: number,
  building: BuildingDef,
): Promise<Record<string, unknown>> {
  const materials = await materialsOf(childId, tx);
  const cost = building.cost;
  const missing: Record<string, number> = {};
  for (const [code, need] of Object.entries(cost)) {
    if ((materials[code] ?? 0) < need) {
      missing[code] = need - (materials[code] ?? 0);
    }
  }
  if (Object.keys(missing).length > 0) {
    return {
      built: false,
      reason: "材料不足",
      materials,
      missing,
      unlocks: [],
    };
  }

  for (const [code, need] of Object.entries(cost)) {
    const rows = await tx
      .select()
      .from(inventory)
      .where(and(eq(inventory.childId, childId), eq(inventory.itemCode, code)));
    const current = rows.length > 0 ? (rows[0]!.count ?? 0) : 0;
    const next = current - need;
    // Python：查不到就 new 一行 count=0 再减 —— upsert 等价
    await tx
      .insert(inventory)
      .values({ childId, itemCode: code, count: next })
      .onConflictDoUpdate({
        target: [inventory.childId, inventory.itemCode],
        set: { count: next },
      });
  }

  const unlockCodesNow: string[] = [];
  for (const unlockCode of building.unlocks ?? []) {
    const exists = await tx
      .select({ unlockCode: unlock.unlockCode })
      .from(unlock)
      .where(and(eq(unlock.childId, childId), eq(unlock.unlockCode, unlockCode)))
      .limit(1);
    if (exists.length === 0) {
      await tx.insert(unlock).values({ childId, unlockCode });
    }
    // 已存在也计入返回列表 —— Python 的 append 在 if 外面，照抄
    unlockCodesNow.push(unlockCode);
  }
  // Python 这里不查重直接 add：重复建造同一建筑会撞主键 —— 两侧行为一致
  await tx.insert(unlock).values({ childId, unlockCode: `building.${building.code}` });
  return {
    built: true,
    materials: await materialsOf(childId, tx),
    unlocks: unlockCodesNow,
  };
}

/**
 * Python `_iso(value)`：aware datetime.isoformat()，naive 才补 "Z"。
 *
 * psycopg2 / node-pg 从 timestamptz 读出的值永远是 aware，所以只实现
 * aware 分支：时区后缀 +00:00（isoformat 对 UTC aware datetime 的输出；
 * psycopg2 的 tzinfo 跟随 session 时区，本地与 Neon 默认都是 UTC）。
 * 微秒：node-pg 只有毫秒精度 —— 毫秒为 0 时两侧 isoformat 都不带小数，
 * 非 0 时 Python 是 6 位、这里是毫秒×1000 补齐 6 位（api-contract 对拍
 * 的时间戳归一化在 #50 统一处理）。
 */
function iso(value: Date): string {
  const pad = (n: number, width = 2) => String(n).padStart(width, "0");
  const base =
    `${pad(value.getUTCFullYear(), 4)}-${pad(value.getUTCMonth() + 1)}-${pad(value.getUTCDate())}` +
    `T${pad(value.getUTCHours())}:${pad(value.getUTCMinutes())}:${pad(value.getUTCSeconds())}`;
  const ms = value.getUTCMilliseconds();
  const fraction = ms === 0 ? "" : `.${pad(ms * 1000, 6)}`;
  return `${base}${fraction}+00:00`;
}

File: backend/service/growth.py
"""世界 / 实验室 / 成长域。

**P2 阶段的定位**：契约里这些端点必须存在且形状正确，但它们的游戏规则
（材料掉落、建筑解锁、徽章、布局）属于 P5 成长体系。因此本模块只实现
"从学习事实中可推导"的部分，规则常量集中放在文件顶部，方便 P5 整体替换。

实验室（§7）不在本模块维护规则：解锁判定统一走 `backend/lab` +
`config/lab/v0.yaml`，这里只提供 `level_of`（派生等级，ADR-0002）。

推导原则：游戏状态永远不是学习状态的权威（见 db/migrations 的注释）。
这里的材料 / 建筑 / 徽章只读 inventory / unlock / attempt / proficiency_state。
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from backend.content.loader import ContentBundle
from backend.db import models
from backend.engine.config import AlgorithmConfig
from backend.engine.graph import CompetencyGraph
from backend.engine.state_machine import derive_level, is_mastered
from backend.engine.types import ChildLearningState
from backend.lab import list_experiments

# ── P5 待接管的常量（内容是资产，规则应可配置） ─────────────
UNIVERSES = [
    {"code": "number_station", "name": "数字车站", "emoji": "🚂"},
    {"code": "detective", "name": "侦探社", "emoji": "🦊"},
]

BUILDINGS = [
    {
        "code": "ticket_booth",
        "name": "售票亭",
        "emoji": "🎫",
        "cost": {"wood": 5, "coin": 2},
        "unlocks": [],
    },
    {
        "code": "platform",
        "name": "站台",
        "emoji": "🛤️",
        "cost": {"wood": 10, "coin": 5},
        "unlocks": ["universe.detective"],
    },
]

BADGES = [
    {"code": "first_make_ten", "name": "第一次凑十", "emoji": "🎖️", "competency": "make_ten"},
    {"code": "first_carry", "name": "第一次进位", "emoji": "🚃", "competency": "carry_add"},
]

MATERIAL_CODES = ["wood", "coin", "gem", "seed"]

COMPETENCY_EMOJI = {
    "sd_add_10": "➕",
    "sd_sub_10": "➖",
    "sd_add_20": "🔢",
    "sd_sub_20": "🔽",
    "make_ten": "🔟",
    "place_value": "🚃",
    "td_add_nocarry": "🧮",
    "td_sub_nocarry": "📉",
    "carry_add": "🔁",
    "borrow_sub": "🔄",
}

DEFAULT_UNIVERSE = "number_station"


# ── 基础读取 ───────────────────────────────────────────────
def materials_of(session, child_id: int) -> Dict[str, int]:
    out = {code: 0 for code in MATERIAL_CODES}
    for row in (
        session.query(models.Inventory)
        .filter(models.Inventory.child_id == child_id)
        .all()
    ):
        out[row.item_code] = int(row.count or 0)
    return out


def unlocked_codes(session, child_id: int) -> set:
    return {
        row.unlock_code
        for row in session.query(models.Unlock)
        .filter(models.Unlock.child_id == child_id)
        .all()
    }


def competency_emoji(code: str) -> str:
    return COMPETENCY_EMOJI.get(code, "📘")


def lab_payload(state: ChildLearningState, cfg: AlgorithmConfig) -> Dict[str, Any]:
    """契约 §7：实验列表。

    解锁规则整套由 `backend/lab` 负责（配置在 `config/lab/v0.yaml`）：
    按能力的**派生等级**（ADR-0002，读时现算，不落库）判定，而不是年级/天数。
    """

    def level_of(competency_code: str) -> str:
        signals = state.competencies.get(competency_code)
        if signals is None:
            return cfg.level_order[0]
        return derive_level(signals, cfg)

    return {
        "experiments": list_experiments(
            level_of=level_of, level_order=cfg.level_order
        )
    }


# ── 故事完成度（用 attempt 的 slot_code 反推） ─────────────
def completed_story_codes(
    session, child_id: int, bundle: ContentBundle
) -> List[str]:
    done: List[str] = []
    for story in bundle.stories.values():
        slot_codes = [
            beat.slot_code for beat in story.challenge_beats() if beat.slot_code
        ]
        if not slot_codes:
            continue
        hit = {
            row[0]
            for row in session.query(models.Attempt.slot_code)
            .filter(
                models.Attempt.child_id == child_id,
                models.Attempt.slot_code.in_(slot_codes),
            )
            .distinct()
            .all()
            if row[0]
        }
        if all(code in hit for code in slot_codes):
            done.append(story.code)
    return done


# ── GET /v1/world ──────────────────────────────────────────
def world_payload(
    session,
    child: models.Child,
    state: ChildLearningState,
    cfg: AlgorithmConfig,
    bundle: ContentBundle,
    budget_minutes: int,
) -> Dict[str, Any]:
    completed = set(completed_story_codes(session, child.id, bundle))
    stories = sorted(bundle.stories.values(), key=lambda s: (s.order_index, s.code))

    universes = []
    for universe in UNIVERSES:
        universe_stories = [s for s in stories if s.universe == universe["code"]]
        total = len(universe_stories)
        progress = (len([s for s in universe_stories if s.code in completed]) / total) if total else 0.0
        universes.append(
            {
                "code": universe["code"],
                "name": universe["name"],
                "emoji": universe["emoji"],
                "unlocked": universe["code"] == DEFAULT_UNIVERSE
                or ("universe.{}".format(universe["code"]) in unlocked_codes(session, child.id)),
                "progress": round(progress, 4),
                "stories": [
                    {
                        "code": story.code,
                        "title": story.title,
                        "completed": story.code in completed,
                        "unlocked": True,
                    }
                    for story in universe_stories
                ],
            }
        )

    today_story = next((s for s in stories if s.code not in completed), None)
    if today_story is not None:
        today = {
            "headline": today_story.title,
            "subtitle": today_story.summary or "去数字车站帮它一把",
            "story_code": today_story.code,
            "universe_code": today_story.universe,
            "estimated_minutes": today_story.duration_min or budget_minutes,
            "completed": False,
        }
    else:
        # 没有故事内容（P2 未接入）时的降级：形状不变，字段指向默认宇宙
        today = {
            "headline": "今天来练一练数字吧",
            "subtitle": "去数字车站练一练",
            "story_code": None,
            "universe_code": DEFAULT_UNIVERSE,
            "estimated_minutes": budget_minutes,
            "completed": False,
        }

    return {
        "child": {"id": child.id, "name": child.name},
        "today": today,
        "universes": universes,
        "lab_unlocked": True,
        "detective_unlocked": "universe.detective" in unlocked_codes(session, child.id),
        "growth_summary": {
            "materials": materials_of(session, child.id),
            "buildings": [
                code.split(".", 1)[1]
                for code in sorted(unlocked_codes(session, child.id))
                if code.startswith("building.")
            ],
            "newest_badge": _newest_badge(session, child.id, state),
        },
    }


# ── GET /v1/growth ─────────────────────────────────────────
def badge_payloads(
    session, child_id: int, state: ChildLearningState
) -> List[Dict[str, Any]]:
    out = []
    for badge in BADGES:
        signals = state.competencies.get(badge["competency"])
        earned = signals is not None and signals.sample_count > 0
        earned_at = None
        if earned:
            row = (
                session.query(models.Attempt.created_at)
                .filter(
                    models.Attempt.child_id == child_id,
                    models.Attempt.competency_code == badge["competency"],
                )
                .order_by(models.Attempt.seq)
                .first()
            )
            if row is not None and row[0] is not None:
                earned_at = _iso(row[0])
        out.append(
            {
                "code": badge["code"],
                "name": badge["name"],
                "emoji": badge["emoji"],
                "earned": earned,
                "earned_at": earned_at,
            }
        )
    return out


def _newest_badge(session, child_id: int, state: ChildLearningState) -> Optional[Dict[str, Any]]:
    earned = [b for b in badge_payloads(session, child_id, state) if b["earned"]]
    if not earned:
        return None
    return {
        "code": earned[-1]["code"],
        "name": earned[-1]["name"],
        "emoji": earned[-1]["emoji"],
    }


def growth_payload(
    session,
    child: models.Child,
    state: ChildLearningState,
    cfg: AlgorithmConfig,
    bundle: ContentBundle,
    graph: CompetencyGraph,
) -> Dict[str, Any]:
    order = graph.topological_order()
    position_index = {code: index for index, code in enumerate(order)}

    nodes = []
    for code in order:
        competency = bundle.competencies.get(code)
        if competency is None:
            continue
        signals = state.competencies.get(code)
        level = derive_level(signals, cfg) if signals is not None else cfg.level_order[0]
        nodes.append(
            {
                "code": code,
                "name": competency.name,
                "level": level,
                "level_label": cfg.level_label(level),
                "mastered": is_mastered(state, code, graph, cfg),
                "emoji": competency_emoji(code),
                "unlocked": True,
                "position": {"x": position_index.get(code, 0), "y": competency.stage},
            }
        )

    edges = []
    for code in order:
        for prereq in graph.prerequisites(code):
            edges.append({"from": prereq, "to": code})

    built = {
        code.split(".", 1)[1]
        for code in unlocked_codes(session, child.id)
        if code.startswith("building.")
    }
    return {
        "tree": {"nodes": nodes, "edges": edges},
        "materials": materials_of(session, child.id),
        "buildings": [
            {
                "code": row["code"],
                "name": row["name"],
                "emoji": row["emoji"],
                "built": row["code"] in built,
                "cost": row["cost"],
            }
            for row in BUILDINGS
        ],
        "badges": badge_payloads(session, child.id, state),
    }


def build_payload(
    session, child_id: int, building: Dict[str, Any]
) -> Dict[str, Any]:
    """POST /v1/growth/build：材料足够则扣减并解锁。"""
    materials = materials_of(session, child_id)
    cost = building["cost"]
    missing = {
        code: need - materials.get(code, 0)
        for code, need in cost.items()
        if materials.get(code, 0) < need
    }
    if missing:
        return {
            "built": False,
            "reason": "材料不足",
            "materials": materials,
            "missing": missing,
            "unlocks": [],
        }

    for code, need in cost.items():
        row = (
            session.query(models.Inventory)
            .filter(
                models.Inventory.child_id == child_id,
                models.Inventory.item_code == code,
            )
            .one_or_none()
        )
        if row is None:
            row = models.Inventory(child_id=child_id, item_code=code, count=0)
        row.count = (row.count or 0) - need
        session.add(row)

    unlock_codes_now = []
    for unlock_code in building.get("unlocks", []):
        exists = (
            session.query(models.Unlock)
            .filter(
                models.Unlock.child_id == child_id,
                models.Unlock.unlock_code == unlock_code,
            )
            .one_or_none()
        )
        if exists is None:
            session.add(models.Unlock(child_id=child_id, unlock_code=unlock_code))
        unlock_codes_now.append(unlock_code)
    session.add(
        models.Unlock(child_id=child_id, unlock_code="building.{}".format(building["code"]))
    )
    session.flush()
    return {
        "built": True,
        "materials": materials_of(session, child_id),
        "unlocks": unlock_codes_now,
    }


def _iso(value) -> str:
    text = value.isoformat()
    if value.tzinfo is None:
        text += "Z"
    return text


__all__ = [
    "UNIVERSES",
    "BUILDINGS",
    "BADGES",
    "materials_of",
    "unlocked_codes",
    "lab_payload",
    "world_payload",
    "growth_payload",
    "build_payload",
    "completed_story_codes",
    "competency_emoji",
]
